src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -268,52 +286,288 @@ internal BoundExpression ConvertPatternExpression(TypeSymbol inputType, CSharpSy
                 {
-                    hasErrors = CheckRestrictedTypeInAsync(this.ContainingMemberOrLambda, declType, diagnostics, typeSyntax);
+                    var location = new SourceLocation(node.SyntaxTree, new Text.TextSpan(node.OpenParenToken.SpanStart, node.CloseParenToken.Span.End - node.OpenParenToken.SpanStart));
+                    diagnostics.Add(ErrorCode.ERR_WrongNumberOfSubpatterns, location, declType.TupleElementTypes, elementTypes.Length, node.SubPatterns.Count);
+                    hasErrors = true;
+                }
+                for (int i = 0; i < node.SubPatterns.Count; i++)
+                {
+                    bool isError = i >= elementTypes.Length;
+                    var elementType = isError ? CreateErrorType() : elementTypes[i];
+                    // PROTOTYPE(patterns2): Check that node.SubPatterns[i].NameColon?.Name corresponds to tuple element i of declType.
+                    var boundSubpattern = BindPattern(node.SubPatterns[i].Pattern, elementType, isError, diagnostics);
+                    patterns.Add(boundSubpattern);
                 }
-
-                return new BoundDeclarationPattern(node, localSymbol, boundDeclType, isVar, hasErrors);
             }
             else
             {
-                // We should have the right binder in the chain for a script or interactive, so we use the field for the pattern.
-                Debug.Assert(node.SyntaxTree.Options.Kind != SourceCodeKind.Regular);
-                GlobalExpressionVariable expressionVariableField = LookupDeclaredField(designation);
-                DiagnosticBag tempDiagnostics = DiagnosticBag.GetInstance();
-                expressionVariableField.SetType(declType, tempDiagnostics);
-                tempDiagnostics.Free();
-                BoundExpression receiver = SynthesizeReceiver(node, expressionVariableField, diagnostics);
-                var variableAccess = new BoundFieldAccess(node, receiver, expressionVariableField, null, hasErrors);
-                return new BoundDeclarationPattern(node, expressionVariableField, variableAccess, boundDeclType, isVar, hasErrors);
+                // It is not a tuple type. Seek an appropriate Deconstruct method.
+
+                var inputPlaceholder = new BoundImplicitReceiver(node, declType); // A fake receiver expression to permit us to reuse binding logic
+                var deconstruct = MakeDeconstructInvocationExpression(
+                    node.SubPatterns.Count, inputPlaceholder, node, diagnostics, out var outPlaceholders, requireTwoOrMoreElements: false);
+                deconstructMethod = deconstruct.ExpressionSymbol as MethodSymbol;
+                // PROTOTYPE(patterns2): Set and check the deconstructMethod
+
+                for (int i = 0; i < node.SubPatterns.Count; i++)
+                {
+                    bool isError = outPlaceholders.IsDefaultOrEmpty || i >= outPlaceholders.Length;
```

Comment:
>i >= outPlaceholders.Length [](start = 71, length = 27)

Can this condition ```i >= outPlaceholders.Length``` ever be true?

Reply:
Yes, in error scenarios.

---
In reply to: [156516950](https://github.com/dotnet/roslyn/pull/23209#discussion_r156516950) [](ancestors = 156516950)